### PR TITLE
fix(toc): exclude time-sensitive promotional headings from TOC extraction

### PR DESCRIPTION
**Pull request by anagarwa**
## Problem
Headings that read as merchandising banners — price drops (`25% Off`), flash sales, limited-time offers, "Save \$50 today" — go stale within days or hours and describe a promo, not a stable content section. A TOC (or an LLM citing it) pointing at "20% Off Everything" next month links to a dead promo.

## Fix
Adds `isTimeSensitivePromoHeadingText()` to `src/headings/utils.js`, wired into `extractTocData()` alongside the existing consent-phrase exclusion (`isExcludedConsentHeadingText`).

Unlike consent-widget copy (a fixed, known set of vendor phrases handled by `TOC_EXCLUDED_HEADING_PHRASES`), promo copy is open-ended — any percentage, dollar amount, or campaign name — so this is a **regex-driven** check rather than a literal phrase list.

Matches:
- `%`-off / `$`-off / "save \$X" phrasing
- Named sale events (flash/clearance/blowout/closeout sale, Black Friday, Cyber Monday)
- Urgency/expiry phrasing ("limited-time offer", "sale ends", "offer expires", "while supplies last", "today only")
- Promo/coupon/discount code call-outs

Deliberately does **not** match bare `"Price"`, `"Pricing"`, `"Sale"`, `"Offer"`, or `"Deals"` alone — those are legitimate evergreen headings (SaaS pricing pages, real-estate "Homes for Sale", a services page's "Our Offers"). Only matches when paired with a quantity/urgency signal.

## Testing
- Added unit tests for `isTimeSensitivePromoHeadingText` (positive cases per pattern category, case-insensitivity, and negative cases for evergreen pricing/offer headings).
- Added `extractTocData()` integration tests: one excludes promo headings, one confirms evergreen pricing/offer headings are kept.
- `npm run test:spec -- test/audits/toc.test.js` → 210 passing, `src/headings/utils.js` at 100% lines/branches/functions/statements.
- `npx eslint src/headings/utils.js` → clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Change Management

```yaml
cm-assessment: v1
changeType: standard
impact: unnoticeable
risk: minor
changeApprovedBy: ["Dominique Jäggi", "Yaashi Madan"]
rationale: "Regex-based heading filter added to TOC extraction; worst case is a heading wrongly included/excluded, self-corrects on next audit run."
```

### CODE REVIEW

File: src/headings/utils.js [hunk 1]
```diff
@@ -76,6 +76,21 @@ export const TOC_EXCLUDED_HEADING_PHRASES = [
...
+ * content section — a TOC (or an LLM citing it) pointing at "20% Off Everything" next month
+ * links to a dead promo. Unlike consent-widget copy above (a fixed, known set of vendor
+ * phrases), promo copy is open-ended — any percentage, dollar amount, or campaign name — so
+ * this is regex-driven instead of a literal phrase list.
+ *
+ * Deliberately does NOT match bare "Price", "Pricing", "Sale", "Offer", or "Deals" alone —
+ * those are legitimate evergreen headings (SaaS pricing pages, real-estate "Homes for Sale",
+ * a services page's "Our Offers"). Only matches when paired with a quantity/urgency signal
+ * (a %, a $ amount, a named sale event, or "ends/expires/today only/while supplies last").
+ */
```
Comment (MysticatBot) [inline src/headings/utils.js:95]: issue (blocking): The `save` branch makes the dollar sign optional, so any heading containing "save" followed by a bare number will match. Headings like "Save 5 Minutes a Day", "Save 3 Steps in Your Workflow", or "Save 100 Trees This Year" are legitimate evergreen headings that would be silently excluded from the TOC.

The code comment says the regex "only matches when paired with a quantity/urgency signal (a %, a $ amount, ...)", but a bare number is neither.

**Fix:** Make `$` mandatory when there is no "up to" prefix. Split into two sub-patterns:
```
save up to $?digits | save $digits
```
This preserves "Save up to 200" and "Save $50" but stops matching "Save 5 Minutes".

**Missing test:** Add a negative assertion for inputs like "Save 5 Minutes a Day" to lock in the intended behavior.

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @anagarwa,

**Verdict: Request changes** - one false-positive gap in the regex will silently drop legitimate headings from the TOC.
**Complexity: LOW** - small, single-service change with good test coverage.
_Changes: Adds a regex-driven filter to exclude time-sensitive promotional headings (price drops, flash sales, limited-time offers) from TOC extraction (2 files)._
Note: CI checks are currently pending - verify build passes before merge.

## Must fix before merge
1. [Important] `save + bare number` false positive silently excludes non-promotional headings - `src/headings/utils.js:91` (details inline)

<details>
<summary>Non-blocking (6): minor issues and suggestions</summary>

- nit: Dollar-amount branch requires exactly two decimal places when present - a heading like "$9.9 Off" would not match. Consider changing to one-or-two decimal digits - `src/headings/utils.js:91`
- nit: First regex branch lacks the leading word boundary that every other branch has. While digits at a non-digit boundary are effectively anchored, adding the anchor would make the pattern consistent - `src/headings/utils.js:91`
- suggestion: Document that the dollar-specific branches only recognize USD amounts. Non-dollar flat discounts (GBP 50 off, EUR 50 off) are not matched. This is an acceptable scope limitation but should be noted in the code comment - `src/headings/utils.js:79-91`
- suggestion: The 460-character single-line regex is hard to review and extend. Consider breaking it into a named array of sub-patterns joined at runtime, similar to how the excluded heading phrases array is structured - `src/headings/utils.js:91`
- nit: The normalize function is called once in the new filter and again in extractTocData at line 362, so the same text gets normalized twice when both filters run. Not a performance concern at realistic heading counts, but worth noting if the filter chain grows - `src/headings/utils.js:129`
- suggestion: The extractTocData JSDoc now lists three exclusion categories in paragraph form. Consider restructuring as a bulleted list for readability as the exclusion set grows - `src/headings/utils.js:329-334`
</details>


---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 0m 13s | <b>Cost:</b> $7.95 | <b>Commit:</b> <code>fd1901d1869f6113f440a3be4a24e0ed8545e324</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: src/headings/utils.js (hunk 1 above)
Comment (anagarwa) [reply src/headings/utils.js:95]: Good catch, fixed in 4e8331d2. Split the `save` alternative so `$` is only optional after "up to" — bare `save N` now requires the literal `$`:

```
\bsave\s+up\s+to\s+\$?\d+\b|\bsave\s+\$\d+\b
```

"Save $50" and "Save up to $200"/"Save up to 200" still match; "Save 5 Minutes a Day", "Save 3 Steps in Your Workflow", "Save 100 Trees This Year" no longer do. Added the negative test you asked for.

**Review by MysticatBot — APPROVED** [review]
@/tmp/mysticat-workspace-kry9jqyl/review-body-2954.md

**Review by MysticatBot — APPROVED** [review]
Hey @anagarwa,

**Verdict: Approve** - the prior blocking finding was addressed correctly; remaining items are non-blocking nits.
**Complexity: LOW** - small, single-service change with good test coverage.
_Changes: Adds a regex-driven filter to exclude time-sensitive promotional headings from TOC extraction, with the save-pattern false-positive fix applied (2 files)._

<details>
<summary>Non-blocking (5): minor issues and suggestions</summary>

- nit: First regex branch `\d+%\s*off\b` lacks a leading `\b` anchor, unlike every other branch. "100% Off-Grid Living" would match since `-` after "off" is a word boundary. One-character fix: `\b\d+%\s*off\b` - `src/headings/utils.js:96`
- nit: Dollar-amount branch `\$\d+(?:\.\d{2})?` accepts exactly two decimal places or none. "$9.9 Off" would not match. Relax to `(?:\.\d{1,2})?` - `src/headings/utils.js:96`
- nit: A few regex alternatives lack test coverage: the `$savings` variant (only `off` is tested), `blowout sale`/`closeout sale`, and the hyphenated form `limited-time` (only the space-separated form is tested) - `test/audits/toc.test.js`
- suggestion: Heading text is normalized three separate times in the `extractTocData` filter path (once per exclusion check, once for the `seen` set). Normalize once at the top of the callback and pass the result through - `src/headings/utils.js:363-375`
- suggestion: The regex is approximately 480 characters on a single line. Consider building from a named array of sub-patterns joined with `|`, which would make diffs clearer when a single alternative changes - `src/headings/utils.js:96`
</details>

## Previously flagged, now resolved
- `save + bare number` false positive: regex split so `$` is only optional after "up to"; negative tests added for "Save 5 Minutes a Day" and similar non-promotional headings.


---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 5m 1s | <b>Cost:</b> $5.58 | <b>Commit:</b> <code>4e8331d2ce80bc619f3a169bec8652485f74e30f</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by ymadobe — APPROVED** [review]
